gsl/span

```diff
@@ -386,6 +387,11 @@ public:
     {
     }
 
+    template<class ArrayElementType = std::add_pointer<element_type>>
+    constexpr span(const std::unique_ptr<ArrayElementType>& ptr, index_type count) : storage_(ptr.get(), count) {}
+
+    constexpr span(const std::shared_ptr<ElementType>& ptr, index_type count) : storage_(ptr.get(), count) {}
```

Comment:
Until `shared_ptr` can officially take multiple elements, I'd be tempted to just have this constructor take a `shared_ptr` and construct a single-element `span` from it. If we hear that a lot of people go to the effort of putting dynamically-allocated arrays into `shared_ptr`, then we can always add the second overload.